### PR TITLE
feat(cli): add custom commands support

### PR DESCRIPTION
**Pull request by alexandrevilain**
## Context

Implement custom commands feature allowing users to create reusable slash commands from markdown files stored in global (~/.kilocode/commands/) or project-specific (.kilocode/commands/) directories. Commands support argument substitution, automatic mode/model switching, and YAML frontmatter configuration.

## Implementation

Users drop markdown files in ~/.kilocode/commands/ or .kilocode/commands/, and those become slash commands. That's it.

### Commands loading steps:
  - On startup, I scan both directories for .md files
  - Parse each file with gray-matter to extract YAML frontmatter
  - Project-level commands override global ones (so you can customize per-project)

### The markdown format:
```md
---
description: What the command does
arguments: [arg1, arg2]
mode: code
model: opus
---

Your prompt template here with $1 and $2 for arguments
```

### When you run the command:
  1. Replace $1, $2, etc. with the actual arguments you passed
  2. Switch to the specified mode/model if provided
  3. Send the processed text as a new chat message

### Decisions

**Mardown**: Same as opencode, claude code, simple to write.

**Directories**: We already have this pattern for skills (~/.kilocode/skills/ and .kilocode/skills/), so just reused the same convention. Users already understand it.

**Arguments substitution**: Tried to keep it dumb - just replace $1 with the first arg, $ARGUMENTS with all args. No fancy templating engine needed.

**Error handling:** If a directory doesn't exist, fine. If model switching fails, warn but keep going. Don't break the whole feature because of one bad file.

## Screenshots

Here is an example with the following custom command:

```md
---
description: Example command that switches mode, model and parses arguments.
mode: ask
model: x-ai/grok-code-fast-1
arguments:
  - what
---

Only say "$1"
```

https://github.com/user-attachments/assets/c0de1f02-0227-4102-9315-6f3362076df6

## How to Test

Create custom commands in `~/.kilocode/commands/`or `.kilocode/commands/` and run `kilo`.
You can use them and show commands help

## Get in Touch

Discord: `alexandrevilain.`

### CODE REVIEW

File: cli/src/commands/custom.ts [hunk 1]
```diff
@@ -0,0 +1,168 @@
...
+				if (mode) command.mode = mode
+				if (model) command.model = model
+
+				commands.set(commandName, command)
+			} catch (error) {
+				logs.warn(`Failed to parse custom command file: ${filePath}`, "CustomCommand", { error })
+			}
+		}
+	} catch {
+		// Directory doesn't exist - that's fine.
```
Comment (marius-kilocode) [inline cli/src/commands/custom.ts]:
```suggestion
	const code = (error as NodeJS.ErrnoException)?.code
    if (code !== 'ENOENT') {
        logs.warn(`Failed to scan command directory: ${dirPath}`, "CustomCommand", { error })
    }
```

File: cli/src/commands/custom.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+	await scanCommandDirectory(projectDir, commands)
+
+	return Array.from(commands.values())
+}
+
+/**
+ * Initialize custom commands from markdown files
+ * Call this after built-in commands are initialized
+ */
+export async function initializeCustomCommands(cwd: string): Promise<void> {
```
Comment (marius-kilocode) [inline cli/src/commands/custom.ts:171]: Should we not check if the command already exists instead of overwriting?

File: cli/src/commands/__tests__/custom.test.ts [hunk 3]
```diff
@@ -0,0 +1,484 @@
...
+			const args = ["arg1", "arg2"]
+
+			const result = substituteArguments(content, args)
+
+			expect(result).toBe("Plain text content")
+		})
+	})
+
+	describe("getCustomCommands", () => {
+		const mockCwd = "/mock/project"
```
Comment (marius-kilocode) [inline cli/src/commands/__tests__/custom.test.ts:174]: These tests might fail on Windows

File: apps/kilocode-docs/docs/cli.md [hunk 4]
```diff
@@ -141,6 +141,142 @@ There are community efforts to build and share agent skills. Some resources incl
...
+- **Project commands**: `.kilocode/commands/` (project-specific)
+
+Commands are simple markdown files with YAML frontmatter for configuration.
+
+### Creating a Custom Command
+
+1. Create the commands directory:
+
+    ```bash
+    mkdir -p ~/.kilocode/commands
```
Comment (marius-kilocode) [inline apps/kilocode-docs/docs/cli.md]: What about Windows?

File: cli/src/commands/custom.ts [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+
+/**
+ * Substitute arguments in command content
+ * Supports: $ARGUMENTS (all args), $1, $2, $3, etc. (positional args)
+ */
+export function substituteArguments(content: string, args: string[]): string {
+	let result = content.replace(/\$ARGUMENTS/g, args.join(" "))
+
+	args.forEach((arg, index) => {
+		result = result.replace(new RegExp(`\\$${index + 1}\\b`, "g"), arg)
```
Comment (marius-kilocode) [inline cli/src/commands/custom.ts]: What about edge cases like `/cmd "Price is $100"`?

File: cli/src/commands/custom.ts [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+}
+
+async function scanCommandDirectory(dirPath: string, commands: Map<string, CustomCommand>): Promise<void> {
+	try {
+		const entries = await fs.readdir(dirPath, { withFileTypes: true })
+
+		for (const entry of entries) {
+			if (!entry.isFile() || !entry.name.toLowerCase().endsWith(".md")) continue
+
+			const commandName = entry.name.slice(0, -3)
```
Comment (marius-kilocode) [inline cli/src/commands/custom.ts:41]: Should we validate the command name? `--test.md` should not be possible right?

**Comment by marius-kilocode** [comment]
Cool feature @alexandrevilain! We just need to get those edge cases right.

**Comment by alexandrevilain** [comment]
Thanks for the review @marius-kilocode ! 

I've addressed all your comments except for the failing tests on Windows. Since I don't have a Windows development environment, I can't confirm whether these tests are actually failing or not. However, given that the tests use mocked implementations for `os.homedir()` and `fs.readdir()`, they should theoretically work cross-platform.

Feel free to let me know what kind of issues you're anticipating on Windows with this approach, and I'll be happy to adjust accordingly!

File: cli/src/commands/custom.ts [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+function isValidCommandName(name: string): boolean {
+	return /^[a-zA-Z0-9][a-zA-Z0-9-]*$/.test(name)
+}
+
+async function scanCommandDirectory(dirPath: string, commands: Map<string, CustomCommand>): Promise<void> {
+	try {
+		const entries = await fs.readdir(dirPath, { withFileTypes: true })
+
+		for (const entry of entries) {
+			if (!entry.isFile() || !entry.name.toLowerCase().endsWith(".md")) continue
```
Comment (marius-kilocode) [inline cli/src/commands/custom.ts:39]: @alexandrevilain I think you misunderstood me here. The `.md` file was just an example. We need to validate here if the format for a custom slash command is correct. Maybe a regex would make sense.

Good:
/review
/review-this

Bad:
/this.is.bad
/file.txt
/$#*()%*$#)

Comment (alexandrevilain) [reply cli/src/commands/custom.ts:39]: @marius-kilocode this was the purpose of `isValidCommandName`. Isn't it sufficient ? 

Maybe I can rework it to become `isValidCommandFilename` and validate against `^[a-zA-Z0-9][a-zA-Z0-9-]*\.md$`. WDYT ?

Comment (marius-kilocode) [reply cli/src/commands/custom.ts:39]: I think I misread the regex. I also tested it one more time. The change looks good to go. Thanks a lot!

**Review by marius-kilocode — APPROVED** [review]

**Comment by marius-kilocode** [comment]
Thanks for the contribution @alexandrevilain!
